parity-clib/parity.hpp

```diff
@@ -0,0 +1,228 @@
+#else
+  if (ud)
+    delete static_cast<parity_rpc_callback *>(ud);
+#endif
+};
+
+typedef std::unique_ptr<::parity_subscription,
+                        decltype(parity_unsubscribe_ws) *>
+    parity_subscription;
+class ParityEthereum final {
+  struct ::parity_ethereum *parity_ethereum_instance;
+  parity_rpc_callback *callback;
+
+public:
+  ParityEthereum(const ParityEthereum &other) = delete;
+  ParityEthereum &operator=(const ParityEthereum &other) = delete;
+  ParityEthereum(ParityConfig config, ParityLogger logger,
+                 parity_rpc_callback new_chain_spec_callback)
+      : parity_ethereum_instance(nullptr),
+        callback(new parity_rpc_callback(new_chain_spec_callback)) {
+    struct ::ParityParams params = {
+        config.config,
+        parity_subscription_callback,
+        callback,
+        parity_destructor_callback,
+        logger.logger,
+    };
+    logger.logger = nullptr;
+    config.config = nullptr;
+    if (parity_start(&params, &this->parity_ethereum_instance))
+      throw std::runtime_error("Failed to start Parity Ethereum");
+  }
+  ParityEthereum(ParityEthereum &&other) noexcept
+      : parity_ethereum_instance(nullptr), callback(nullptr) {
+    this->parity_ethereum_instance = other.parity_ethereum_instance;
+    other.parity_ethereum_instance = nullptr;
+  }
+  ParityEthereum &operator=(ParityEthereum &&other) noexcept {
+    if (this != &other) {
+      this->parity_ethereum_instance = other.parity_ethereum_instance;
+      other.parity_ethereum_instance = nullptr;
+    }
+    return *this;
+  }
+  ~ParityEthereum() { parity_destroy(parity_ethereum_instance); }
+
+  /// Perform an asychronous RPC request in a background thread.
+  ///
+  /// @param callback Callback to be called on a background thread.
+  /// This must not throw an exception ― if it does, `std::terminate` is called.
+  /// The callback’s destructor not called, and sizeof(callback) heap space is
+  /// leaked.  This is a bug and will be fixed.  Note that when it is fixed, the
+  /// destructor will be called on an arbitrary thread.
+  void
+  rpc(const std::string_view rpc_query, const std::size_t timeout_ms,
+      parity_rpc_callback &callback) const {
+    typedef std::remove_reference<decltype(callback)>::type callback_ptr;
+    callback_ptr copied_callback = callback;
+    if (::parity_rpc(this->parity_ethereum_instance, rpc_query.data(),
+                     rpc_query.size(), timeout_ms, parity_cplusplus_callback,
+                     parity_destructor_callback,
+                     new callback_ptr(copied_callback)))
+      throw std::runtime_error("Parity RPC failed");
+  }
+  parity_subscription subscribe(
+      const std::string_view buffer,
+      parity_rpc_callback callback) const {
+    if (::parity_subscription *session = ::parity_subscribe_ws(
+            this->parity_ethereum_instance, buffer.data(), buffer.size(),
+            parity_subscription_callback,
+            new parity_rpc_callback(callback),
```

Comment:
If `parity_subscribe_ws` returns null, you have a memory leak here.

Reply:
Fixed by using explicit destructor callbacks.